### PR TITLE
[tpm] rework emulation tpm to use random seed

### PR DESCRIPTION
**Pull request by Itxaka**
There is no point on having a static seed for the registration when working with emualted tpm. In fact, once the registration is done the seed value is never used again so why not just use a random value onr egister and drop that config?

This makes several machines with emulated tpm work with just one registration+iso and has no downsides that I can see

Signed-off-by: Itxaka <igarcia@suse.com>

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -47,11 +47,10 @@ func (in *Install) DeepCopyInto(out *Install) {
 }
 
 type Registration struct {
-	URL             string `json:"url,omitempty" yaml:"url,omitempty" mapstructure:"url"`
-	CACert          string `json:"ca-cert,omitempty" yaml:"ca-cert,omitempty" mapstructure:"ca-cert"`
-	EmulateTPM      bool   `json:"emulate-tpm,omitempty" yaml:"emulate-tpm,omitempty" mapstructure:"emulate-tpm"`
-	EmulatedTPMSeed int64  `json:"emulated-tpm-seed,omitempty" yaml:"emulated-tpm-seed,omitempty" mapstructure:"emulated-tpm-seed"`
```
Comment (alexander-demicev) [inline pkg/config/config.go:53]: this structure is nested in the machine registration object, removing the field makes it a breaking change. It's only used for development so this is probably fine but we have to be aware that such change removes a field from a released API.

Comment (Itxaka) [reply pkg/config/config.go:53]: was thinking about setting here the special value `-1` to indicate "create a random seed" and then thought...why not just create it all the time? Why do we have a explicit seed in here? whats the purpose?

So yeah, sent this to run the tests and check, we can still do what I thought on my first implementation, just check for the `-1` and then generate a random seed, while keeping the struct intact, but I dont think it has any value in there :/

Comment (Itxaka) [reply pkg/config/config.go:53]: We even mention on the docs that this is for debugging purpouses only. In fact, re-reading it, it appears as we use a random seed and you cant then set that value to use a fixed one...yikes, not good

`fixed seed to use with 'emulate-tpm': use for debug purposes only`

**Comment by kkaempf** [comment]
I'd veto dropping the `emulated-tpm-seed`.
This is helpful in testing (a pre-populated machineInventory) and will be essential when implementing/testing secure device onboarding.

**Comment by Itxaka** [comment]
@kkaempf nicer? This only kicks in if seed its set to -1, leaving the normal seed stuff as is

**Comment by kkaempf** [comment]
> @kkaempf nicer? This only kicks in if seed its set to -1, leaving the normal seed stuff as is

💚

**Review by fgiudici — COMMENTED** [review]
The PR looks good! The linter issue should be already fixed on the controller_runtime branch that will become our new main... so, probably we may port this branch on the controller_runtime and have the linter happy :-) 

Apart from the PR itself, I'm worried about supporting TPM emulation: this defeats completely using TPM for registering the hosts, i.e., to let the joining machine to proof its identity. With emulation, anyone with the seed can "impersonate" the host. This is why I think was set as a debug feature only.
So, not against merging this at all, but let's state at the very least this is not  a supported configuration for production (what about at least a warning message?). Easy anyway to forget...
If we want to allow elemental nodes without TPM, I would add other auth methods. But this is for another discussion... ;-)

**Comment by Itxaka** [comment]
> Apart from the PR itself, I'm worried about supporting TPM emulation: this defeats completely using TPM for registering the hosts, i.e., to let the joining machine to proof its identity. With emulation, anyone with the seed can "impersonate" the host. This is why I think was set as a debug feature only.
> So, not against merging this at all, but let's state at the very least this is not a supported configuration for production (what about at least a warning message?). Easy anyway to forget...
> If we want to allow elemental nodes without TPM, I would add other auth methods. But this is for another discussion... ;-)

Agreed but this is kind of out of our control now that we released with that possibility. I think its nice to have and if someone wants to deploy without proper TPM then its their own responsability, same as if they disable secureboot. We can recommend enabling it and write down that its not secure to have emulated TPM but then the ball is in the customer field.

Still, its a good thing we had emulated TPM, as there seems to be a lot of people testing this on non-TPM-capable environments. At least they can try/demo that :)

I almost forgot! cc @abonillabeeche you will probably like this ;)

**Comment by agracey** [comment]
> this defeats completely using TPM for registering the hosts, i.e., to let the joining machine to proof its identity

If we merge this*, let's add an issue about being able to distinguish where the uniqueness comes from in the MachineInventory.

* I don't completely disagree with this feature but it needs to be clearly documented as "developer only and not to be used in production environments"

File: pkg/config/config.go (hunk 1 above)
Comment (agracey) [reply pkg/config/config.go]: Just to make sure, this seed (or the emulated tpm hash) will get persisted across reboots and upgrades, right?

Comment (Itxaka) [reply pkg/config/config.go]: I think the TPMHash derived from this is the one used afterwards...Im not even sure its used afterwards, only in the registration? 

@fgiudici I think you have the most knowledge in this area, is the TPMHash used afterwards for secure communication, or its just used on the first registration only? I cant see it used anywhere else....

**Comment by Itxaka** [comment]
> > this defeats completely using TPM for registering the hosts, i.e., to let the joining machine to proof its identity
> 
> If we merge this*, let's add an issue about being able to distinguish where the uniqueness comes from in the MachineInventory.
> 
> * I don't completely disagree with this feature but it needs to be clearly documented as "developer only and not to be used in production environments"

Agreed. Note that this PR is not the one that defeats using TPM for auth, its the original tpm emulation the one that disregards using TPM for unique auth. In fact I would say that having a random seed makes it more secure as you cannot "clone" other machines in the same deployment easily without knowing the seed, while having a static seed of 1 makes it trivial to reproduce the machine TPM hash.

In fact I would really like to enable this value (-1) as the default so the seed is always random and if you want a static seed for whatever testing, you have to go an enable it specifically.

In any case, even if this is not merged, the docs need updating to make sure we have a specific section called debug or developer in which we mention both the emulate tpm and the tpm seed on its own, instead of having it as part of the default config docs, so its clear its not something normal or expected, but something for developing or debugging.

File: pkg/config/config.go (hunk 1 above)
Comment (Itxaka) [reply pkg/config/config.go]: FYI I deployed a cluster using the random seed mechanism already and could not see any issues, hence me wondering where is the TPMSeed used afterwards. system-agent nor k8s has any concept of the TPMHash so I think its only valid for the registration.

Comment (kkaempf) [reply pkg/config/config.go]: I use it (soft-tpm with a fixed seed) regularly to test "create machineInventory first, then run installation" with RPi.

Comment (Itxaka) [reply pkg/config/config.go]: yes, but after registration Im not sure where its used. Also that same workflow would work with a random seed, I cant see how would having a fixed VS random seed would give us any advantage here but maybe its becasue I dont know that much about the tpm stuff in the registration

**Review by fgiudici — CHANGES_REQUESTED** [review]
I took some time to re-look at the code. It really generates random keys for the TPM simulator. You can disable the random keys generation and have all derived by using a specific seed, making it insecure. But, by default, with a seed of 0, keys are generated randomly and are secure. Once the registration process is over anyway, the keys are lost and the host will not be able to re-register to update labels or whatever (but this is not something we rely or use anymore).
Guess it is the reason we CHANGED THE DEFAULT EMULATED-TPM-SEED to 1 (we weren't able to update the labels in the very old version of the operator) 🤦🏼  

So, long story short, if you set the tpm-seed to 0 it is already randomized, if you don't put the seed it gets 1 and it is NOT randomized.
We may want just too fix the default at: https://github.com/rancher/elemental-operator/blob/v1.0.2/cmd/register/main.go#L117

**Comment by Itxaka** [comment]
> I took some time to re-look at the code. It really generates random keys for the TPM simulator. You can disable the random keys generation and have all derived by using a specific seed, making it insecure. But, by default, with a seed of 0, keys are generated randomly and are secure. Once the registration process is over anyway, the keys are lost and the host will not be able to re-register to update labels or whatever (but this is not something we rely or use anymore). Guess it is the reason we CHANGED THE DEFAULT EMULATED-TPM-SEED to 1 (we weren't able to update the labels in the very old version of the operator) 🤦🏼
> 
> So, long story short, if you set the tpm-seed to 0 it is already randomized, if you don't put the seed it gets 1 and it is NOT randomized. We may want just too fix the default at: [`v1.0.2`/cmd/register/main.go#L117](https://github.com/rancher/elemental-operator/blob/v1.0.2/cmd/register/main.go?rgh-link-date=2022-12-12T16%3A31%3A32Z#L117)

wat

looool

good find @fgiudici !

**Comment by fgiudici** [comment]
Since the emulated-tpm-seed set to 0 cannot work right now, and following discussion #235 ... reopening this (while running away from @Itxaka 🤣 )

**Review by fgiudici — APPROVED** [review]
